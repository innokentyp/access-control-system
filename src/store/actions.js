import { 
	LOGIN,

  LOGIN_SUCCEEDED, 
  LOGIN_FAILED,

  LOGOUT
} from './constants'

export function login(credentials) {
	return {
		type: LOGIN,
		credentials
	}
}

export function loginSucceeded(user) {
	return {
		type: LOGIN_SUCCEEDED,
		user
	}
}

export function loginFailed(message) {
	return {
		type: LOGIN_FAILED,
		message
	}
}

export function logout() {
	return {
		type: LOGOUT
	}
}