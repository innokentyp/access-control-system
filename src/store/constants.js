export const REQUEST_DATA = Symbol('Request data')

export const LOGGED_IN = Symbol('Logged in')
export const LOGGED_OUT = Symbol('Logged out')

export const REQUEST_PLACES = Symbol('Request places')
export const PLACES_FETCHED = Symbol('Places fetched')

export const ADD_PLACE = Symbol('Add place')
export const UPDATE_PLACE = Symbol('Update place')

export const PLACE_PUTTED = Symbol('Place putted')
export const PLACE_PATCHED = Symbol('Place patched')

export const REQUEST_SUBJECTS = Symbol('Request subjects')
export const SUBJECTS_FETCHED = Symbol('Subjects fetched')

export const REQUEST_SUBJECT = Symbol('Request subject')
export const SUBJECT_FETCHED = Symbol('Subject fetched')