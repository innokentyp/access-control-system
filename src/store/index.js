import { createStore, applyMiddleware } from 'redux'

import thunkMiddleware from 'redux-thunk'
import createSagaMiddleware from 'redux-saga'

import { createLogger } from 'redux-logger'

import reducer, { preloadedState } from './reducers'
import requestSubjectsSaga from './sagas/requestSubjectsSaga'

export default function configureStore() {
	// create the saga middleware
	const sagaMiddleware = createSagaMiddleware()

	const store = createStore(reducer, preloadedState, applyMiddleware(thunkMiddleware, sagaMiddleware, createLogger()))  

	// then run the saga
	sagaMiddleware.run(requestSubjectsSaga)

	// return
	return store
}