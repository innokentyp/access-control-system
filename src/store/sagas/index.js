import { all } from 'redux-saga/effects'

import requestSubjectsSaga from './requestSubjectsSaga'

export default function* rootSaga() {
  yield all(
	  [
	    requestSubjectsSaga()
	  ]
  )
}