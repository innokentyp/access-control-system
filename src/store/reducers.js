import { combineReducers } from 'redux'

import * as types from './constants'

// constants
personal_editor.MAX_SUBJECTS = 12

// preloaded state
export const preloadedState = {
  authentication: {
    user: {}
  },

  personal: {
    subjects: [],

    filtering: '',
    sorting: ''
  },

  personal_editor: {
    subjects: {}
  } 
}

// reducers
function authentication(state = preloadedState.authentication, action) {
  switch (action.type) {
    case types.LOGGED_IN:
      return { ...state, user: { ...action.user } }
    case types.LOGGED_OUT:
      return { ...state, user: {} }  
    default:
      return state 
  }
}

function personal(state = preloadedState.personal, action) {
  switch (action.type) {    
    case types.SUBJECTS_FETCHED:
      return { ...state, subjects: action.subjects }

    case types.SET_FILTERING: 
      return { ...state, filtering: action.filtering }  
    case types.SET_SORTING: 
      return { ...state, sorting: action.sorting }  
    default:
      return state 
  }
}

function personal_editor(state = preloadedState.personal_editor, action) {
  switch (action.type) {    
    case types.SUBJECT_FETCHED:
      const subjects = { ...state.subjects, [action.subject.id]: action.subject }
      const ids = Object.keys(subjects)

      while (ids.length > personal_editor.MAX_SUBJECTS) {
        delete subjects[ids[0]]
        ids.splice(0, 1)
      }

      return { ...state, subjects }    
    default:
      return state 
  }
}

export default combineReducers({ authentication, personal, personal_editor })