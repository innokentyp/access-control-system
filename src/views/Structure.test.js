import React, { Component } from 'react'
import { Provider } from 'react-redux'

import { MemoryRouter, Route } from 'react-router-dom'
import { withRouter } from 'react-router'

import { configure, shallow, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import toJson from 'enzyme-to-json'

import configureMockStore from 'redux-mock-store'

import ConnectedStructure, { Structure } from './Structure'
import { preloadedState } from '../store/reducers'

const middlewares = []
const mockStore = configureMockStore(middlewares)

configure({ adapter: new Adapter() })

global.sessionStorage = {
  getItem: jest.fn(),
  setItem: jest.fn(),
  clear: jest.fn()
}

const history = {
	replace: jest.fn(),
	push: jest.fn()
}

const places = [
	{
    id: '1w18g4bzr5qhv36l',
    name: 'Внешняя территория',
    maximum_control: 0
  },
  {
    id: '1w18g4bzsoe2f6ue',
    name: 'Офис',
    maximum_control: 0,
    places: [
			{
				id: '1w18g4bzwzi6rwmj',
        name: 'Этаж 1',
        maximum_control: 0
			},
			{
				id: '1w18g4bzzjrjqed2',
        name: 'Этаж 2',
        maximum_control: 0
			}
    ]
  }
]

const actions = {
	requestPlaces: jest.fn(),
	addPlace: jest.fn(),
	deletePlace: jest.fn(),
	changePlaceParent: jest.fn(),
	post: jest.fn(), 
	patch: jest.fn(), 
	remove: jest.fn()
}
/*
jest.mock('../store/', 
	() => (
		{
			_store: {
				getState: jest.fn().mockImplementation(
					() => { 
						return {
							structure: {
								places: global.places
							}
						} 
					}	
				)
			}
		}
	)
)
*/
describe('Structure', 
	() => {
		it('renders without crashing',
			() => {
				const props = {
					match: {
						url: '/structure'
					},
					location: { 
						pathname: '/structure'
					},
					history,
					structure: {
				    places,
				    at: new Date(2018, 1, 1).getTime(),    
				    inserted: [], updated: [], deleted: []
				  },
				  actions 
				}

				const wrapper = shallow(<Structure {...props} />)

				expect(toJson(wrapper)).toMatchSnapshot()
			}
		)
		
		it('correct render for deep place',
			() => {
				const state = {
					...JSON.parse(JSON.stringify(preloadedState)),
					structure: {
				    places,
				    at: new Date(2018, 1, 1).getTime(),    
				    inserted: [], updated: [], deleted: []
				  }
				}

				const place = {
					id: '1w18g4bzwzi6rwmj',
					name: 'Этаж 1' 
				}

				const wrapper = mount(
					<Provider store={mockStore(state)}>
						<MemoryRouter initialEntries={[ `/structure/1w18g4bzsoe2f6ue/${place.id}` ]}>
							<Route path='/structure' render={props => <ConnectedStructure {...props} />} />													
						</MemoryRouter>
					</Provider>	
				)

				expect(wrapper.find(`PlaceItem[name="${place.name}"]`)).toHaveLength(1)
				expect(wrapper.find(`PlaceEditor input[name="place-name"][value="${place.name}"]`)).toHaveLength(1)
			}
		)
		
	}
)