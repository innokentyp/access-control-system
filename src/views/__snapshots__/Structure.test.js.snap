// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Structure renders without crashing 1`] = `
<Container
  as="section"
>
  <Header
    as="h4"
  >
    Структура помещений
  </Header>
  <Menu
    secondary={true}
  >
    <Route
      path="/structure/:id"
    />
  </Menu>
  <Grid
    stackable={true}
  >
    <GridRow>
      <GridColumn
        width={6}
      >
        <List>
          <ListItem
            key="1w18g4bzr5qhv36l"
          >
            <ListIcon
              color="grey"
              name="right angle"
              style={
                Object {
                  "visibility": "hidden",
                }
              }
            />
            <ListContent>
              <PlaceItem
                add={[Function]}
                change={[Function]}
                name="Внешняя территория"
                remove={[Function]}
                to="/structure/1w18g4bzr5qhv36l"
              />
            </ListContent>
          </ListItem>
          <ListItem
            key="1w18g4bzsoe2f6ue"
          >
            <ListIcon
              color="grey"
              name="right angle"
              style={
                Object {
                  "visibility": "visible",
                }
              }
            />
            <ListContent>
              <PlaceItem
                add={[Function]}
                change={[Function]}
                name="Офис"
                remove={[Function]}
                to="/structure/1w18g4bzsoe2f6ue"
              />
            </ListContent>
          </ListItem>
          <ListItem>
            <ListIcon
              name="down triangle"
              style={
                Object {
                  "visibility": "hidden",
                }
              }
            />
            <ListContent
              as="a"
              href="/structure"
              onClick={[Function]}
              onDragOver={[Function]}
              onDrop={[Function]}
            >
              ( Новый корневой элемент )
            </ListContent>
          </ListItem>
        </List>
      </GridColumn>
      <GridColumn
        style={
          Object {
            "minHeight": "215px",
          }
        }
        width={10}
      >
        <Route
          component={[Function]}
          path="/structure/:id"
        />
      </GridColumn>
    </GridRow>
    <GridRow>
      <GridColumn>
        <Button
          as="button"
          disabled={true}
          onClick={[Function]}
          positive={true}
          role="button"
        >
          <Icon
            as="i"
            name="check"
          />
           Записать
        </Button>
        <Button
          as="button"
          negative={false}
          onClick={[Function]}
          role="button"
        >
          <Icon
            as="i"
            name="refresh"
          />
           Обновить
        </Button>
      </GridColumn>
    </GridRow>
    <GridRow>
      <GridColumn
        width={5}
      >
        Inserted:
        <pre>
          []
        </pre>
      </GridColumn>
      <GridColumn
        width={5}
      >
        Updated:
        <pre>
          []
        </pre>
      </GridColumn>
      <GridColumn
        width={6}
      >
        Deleted:
        <pre>
          []
        </pre>
      </GridColumn>
    </GridRow>
  </Grid>
</Container>
`;
